### ISSUE TITLE
[ENHANCEMENT] Automatic handling of HeadMoved

### ISSUE THREAD

**Issue by mattyg** [holochain/holochain#3008]
I would like the conductor to handle `HeadMoved` errors without pushing them up to the app developer. Upon receiving a HeadMoved, it should rerun validation on those actions over the new chain top and attempts to write them again.

**Comment by ThetaSinner** [comment]
This would be really useful. A bit like telling Git 'please rebase automatically when I do a pull and both the local and remote have changed'.

I think the tricky part of this would be defining what a conflict looks like when moving to the new chain head. Or maybe this is something the code can already do that I don't know about?

@maackle or @thedavidmeister do you have any thoughts about the complexity of this feature? Does it warrant being a discussion to start with?

**Comment by mattyg** [comment]
> I think the tricky part of this would be defining what a conflict looks like when moving to the new chain head. 

I think I'm not sure what you mean? Wouldn't it just assume no conflict if validation passes again?

**Comment by mattyg** [comment]
> I think what is requested here already exists, in this form: https://docs.rs/hdk/latest/hdk/prelude/enum.ChainTopOrdering.html#variant.Relaxed. Does this do the trick @mattyg ?

I don't think so because I *do* want validation to be re-run on the new chain head because I want to use order-dependent validation. Unless the docs are misaligned with the implementation -- it seems like `ChainTopOrdering::Relaxed` just re-hashes and writes the actions on the new chain head without re running validation.

**Comment by maackle** [comment]
Yes I remember discussions about this a long time ago. I personally feel there should be an option to do what you say @mattyg, and I question the value of even having an option that lets you rebase actions automatically *without* running validation again. I think the Relaxed option should just re-run validation, so that it gives you what you're asking for. If there's any reason to keep the existing Relaxed behavior (probably is one that I'm forgetting), then there should be a third option to do what you say (maybe call the new one "Relaxed" and call the existing one "RelaxedUnvalidated"?)

**Comment by guillemcordoba** [comment]
Oh wow okey I didn't realize that it didn't re-run validation. If that's the case that's a serious security hole I think, I like what you are proposing @maackle but I don't think I would want `RelaxedUnvalidated` to exist, it's one of the biggest foot guns a happ dev could have.

**Comment by mattyg** [comment]
I realized that without this improvement, you're pretty much guaranteed to get chain head moved errors if you use *any* scheduled functions -- either the scheduled function will try to commit and get the error, or an outside zome caller will try to commit and get an error.

At least when all the zome calls are coming from your client app you can manage them coherently to avoid parallel commits -- if you have any scheduled functions running on an interval you can't do that.

This issue is actually a pretty big priority for me.

**Comment by maackle** [comment]
This could be relatively quick to do, especially since we have our heads in the validation workflows, @jost-s @ThetaSinner do you suppose this is a quick item? To just validate records that are being retried during a HeadMoved error in Relaxed mode? I'm not volunteering you for the task but only for the assessment.

**Comment by ThetaSinner** [comment]
I think this is reasonable as a medium piece of work. The data is created and written to the scratch space, then separately flushed to the database and that's where the error would happen. If we wanted the same code to recreate the original data then it's complicated but if this is just a matter of updating sequence numbers and hash pointers then I don't see why it couldn't be done as part of the scratch flush relatively easily.

The only caveat I'm going to throw in here is that I'm pretty sure that there's limited test coverage for this code so if we want to deal with that at the same time it's medium and not small in terms of effort :) 

Happy to work on this one

**Comment by thedavidmeister** [comment]
@ThetaSinner @maackle @mattyg  there's been pretty long conversations about this before in chat, i think @artbrock was the one mostly holding for the not-rerun-validation scenario, but had pretty specific use cases in mind

that enum is there specifically to allow tweaking the behaviour so a with-validation option seems reasonable to me

**Comment by ThetaSinner** [comment]
- For usability, we should make it easier to configure `Relaxed` creates and entries. Would be nice to have a helper.
- The action hash that is returned by the HDK function call will be different from the "rebased" hash. So you have to use the post_commit one. Can we make that more usable?

**Comment by mattyg** [comment]
Another gotcha from docs discussion: https://github.com/holochain/docs-pages/pull/453#discussion_r1863925441

Lets say you have a zome fn that:
- Creates an Entry
- Creates a Link that points to that Entry's ActionHash

You call that zome fn and it fails due to HeadMoved.

If you were to automatically rebase the actions on your source chain and run validation again then:
- The CreateEntry action would have a different action hash
- The CreateLink action would be pointing to the old action hash

Thus validation would fail.

I'm not sure what the ideal solution is.

**Comment by pdaoust** [comment]
I think that in that case it's an indication that the developer's desires -- to reference action hashes created in one transaction from later actions in the same tranasction -- is incompatible with relaxed chain top ordering, and validation is doing exactly what it should do when it fails. Relaxed should probably be for very very simple transactions.

The only workable solution I can imagine is if source-chain writing functions returned some sort of 'promise' of an action hash that can then be used in subsequent actions in the same function. Whether that's even remotely possible, I do not know :)
